kliff/trainer/lightning_trainer.py

```diff
@@ -0,0 +1,514 @@
+
+        TODO:
+            Support for Stresses
+
+        Args:
+            batch: batch to take step over
+            batch_idx: batch index, input from lightning
+
+        Returns:
+            loss: loss for the batch
+        """
+        target_energy = batch.energy
+        target_forces = batch.forces
+
+        energy_weight = self.energy_weight
+        forces_weight = self.forces_weight
+
+        predicted_energy, predicted_forces = self.forward(batch)
+
+        loss = energy_weight * F.mse_loss(
+            predicted_energy.squeeze(), target_energy.squeeze()
+        ) + forces_weight * F.mse_loss(
+            predicted_forces.squeeze(), target_forces.squeeze()
+        )
+        self.log(
+            "train_loss",
+            loss,
+            on_step=False,
+            on_epoch=True,
+            prog_bar=True,
+            logger=True,
+            sync_dist=True,
+        )
+        return loss
+
+    def configure_optimizers(self) -> Union[torch.optim.Optimizer, Dict]:
+        """
+        Configure and return optimizer and learning rate scheduler
+
+        Returns:
+            A dict of optimizer and lr_scheduler if scheduler is requested, else it
+            returns just the optimizer.
+        """
+        optimizer = getattr(torch.optim, self.optimizer_name)(
+            self.model.parameters(), lr=self.lr
+        )
+
+        if self.lr_scheduler:
+            self.lr_scheduler = getattr(torch.optim.lr_scheduler, self.lr_scheduler)(
+                optimizer, **self.lr_scheduler_args
+            )
+            return {
+                "optimizer": optimizer,
+                "lr_scheduler": {
+                    "scheduler": self.lr_scheduler,
+                    "interval": "epoch",
+                    "frequency": 1,
+                    "monitor": "val_loss",
+                },
+            }
+        else:
+            return optimizer
+
+    def validation_step(self, batch, batch_idx) -> Dict[str, torch.Tensor]:
+        """
+        Single validation step for lightning.
+
+        Args:
+            batch: batch to take step over
+            batch_idx: batch index, input from lightning
+
+        Returns:
+            loss: loss for the batch, and per atom force loss for loss trajectory
+        """
+        torch.set_grad_enabled(True)
+
+        target_energy = batch.energy
+        target_forces = batch.forces
+
+        energy_weight = self.energy_weight
+        forces_weight = self.forces_weight
+
+        predicted_energy, predicted_forces = self.forward(batch)
+
+        per_atom_force_loss = torch.sum(
+            (predicted_forces.squeeze() - target_forces.squeeze()) ** 2, dim=1
+        )
+
+        loss = (
+            energy_weight
+            * F.mse_loss(predicted_energy.squeeze(), target_energy.squeeze())
+            + forces_weight * torch.mean(per_atom_force_loss) / 3
+        )  # divide by 3 to get correct MSE
+
+        self.log(
```

Comment:
Besides a sum of the loss, people would typically be more interested in separate losses on energy and forces. 

Also, people can also be interested in metrics like MAE other than MSE.

Most importantly, it is not interesting to know losses/metrics of a validation step for a batch of data. The losses/metrics at each epoch for all data matter. This requires aggregating the data between steps. I'd suggest using torchmetrics, which automatically does it.   

So, we need to report separate losses, and allow other metrics, e.g. using the [trochmetrics](https://lightning.ai/docs/torchmetrics/stable/) package.

Reply:
Torch metrics looks great. I will add support for it. 
Can you explain this a bit:
> Most importantly, it is not interesting to know losses/metrics of a validation step for a batch of data. The losses/metrics at each epoch for all data matter.

Reply:
Currently, the validation loss/metrics is evaluated and reported for each mini batch of data, which is only part of all the validation data. But people generally would be interested in the loss/metrics for all validation/test data, not each mini batch.

Reply:
I dont think so. In logging when we give `on_epoch=True` and `on_step=False`, then I believe the logger will call `after_batch_end` function to simply accumulate the loss and only log it in `after_epoch_end` function. In the log files too I could see that validation losses are logged per epoch, and not per batch. This is similar to what I am doing in the loss traj callback, where `on_validation_batch_end` just gathers the result and `on_validation_epoch_end` writes them to a file. I can recheck the API but I am quite certain that is the case.